### PR TITLE
fix: Redmine 5 compatibility and code cleanup

### PR DESCRIPTION
**Pull request by Hareramrai**
- Set the required Redmine version to minimum 5.0
- before_filter -> before_action
- Fixed auto loading issue & reorganize the code 
- Fixed undefined method issue for alias method chain
- rails 2.x support dropped entirely, removed from code
- Replace `reorder_links` method with `reorder_handle`
- Remove `preview_link` as it's not needed

### CODE REVIEW

**Comment by jdehaan** [comment]
I found an issue if you go to http://localhost:8888/redmine/dev/projects/mars-project-mgmt/settings

```
ActionView::Template::Error
undefined method `reorder_links' for #<ActionView::Base:0x00000000049818>
Did you mean? reorder_handle
/plugins/redmine_boards_watchers/app/views/projects/settings/_boards.html.erb:17:in `block in _plugins_redmine_boards_watchers_app_views_projects_settings__boards_html_erb__1196326932659837139_150520'
/usr/local/bundle/gems/activerecord-6.1.7.2/lib/active_record/relation/delegation.rb:88:in `each'
/usr/local/bundle/gems/activerecord-6.1.7.2/lib/active_record/relation/delegation.rb:88:in `each'
```

File: lib/boards_watchers/patches.rb [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+      end
+    end
+  end
+end
+
+unless Message.included_modules.include? BoardsWatchers::Patches::StickyPriorityMessagePatch
+  Message.include BoardsWatchers::Patches::StickyPriorityMessagePatch
+end
+
+if Rails::VERSION::MAJOR >= 6 && Rails.env.production? && !(ApplicationHelper.included_modules.include? BoardsWatchers::Patches::ApplicationHelperPatch)
```
Comment (jdehaan) [inline lib/boards_watchers/patches.rb]:
```suggestion
if Rails.env.production? && !(ApplicationHelper.included_modules.include? BoardsWatchers::Patches::ApplicationHelperPatch)
```

File: lib/boards_watchers/patches.rb [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+  ApplicationHelper.include BoardsWatchers::Patches::ApplicationHelperPatch
+end
+
+if Redmine::VERSION::MAJOR >= 5 && !(MessagesController.included_modules.include? BoardsWatchers::Patches::MessagesControllerPatch)
```
Comment (jdehaan) [inline lib/boards_watchers/patches.rb]:
```suggestion
if !(MessagesController.included_modules.include? BoardsWatchers::Patches::MessagesControllerPatch)
```

File: lib/boards_watchers_hooks.rb [hunk 3]
```diff
@@ -1,86 +1,88 @@
...
+    end
+    if options.delete(:disabled)
+      options.delete(:method)
+      options.delete(:confirm)
+      options.delete(:onclick)
+      options[:class] << ' disabled'
+      url = '#'
+    end
+
+    if url.is_a?(Hash) && Rails::VERSION::MAJOR >= 3 && Redmine::Utils.relative_url_root != ''
```
Comment (jdehaan) [inline lib/boards_watchers_hooks.rb]:
```suggestion
    if url.is_a?(Hash) && Redmine::Utils.relative_url_root != ''
```

File: lib/boards_watchers/patches.rb [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+      base.send(:include, InstanceMethods)
+      base.class_eval do
+        unloadable
+
+        if (Redmine::VERSION.to_a[0..2] <=> [1, 3, 3]) < 0
+          attr_protected :sticky_priority
+        else
+          safe_attributes 'sticky_priority',
+                          if: ->(message, user) { user.allowed_to?(:edit_messages, message.project) }
+        end
```
Comment (jdehaan) [inline lib/boards_watchers/patches.rb]:
```suggestion
        safe_attributes 'sticky_priority',
                        if: ->(message, user) { user.allowed_to?(:edit_messages, message.project) }
```

**Comment by Hareramrai** [comment]
@jdehaan, thanks for reviewing the PR. I have fixed the issue & it's up code review again.

**Comment by jdehaan** [comment]
Looks much better now. I had to dig deeper in the functionality of the plugin to detect another crash:

http://localhost:8888/redmine/dev/projects/mars-project-mgmt/boards/31

```
ActionView::Template::Error
undefined method `preview_link' for #<ActionView::Base:0x00000000046320>
Did you mean? preview_news_url
/plugins/redmine_boards_watchers/app/views/boards/show.html.erb:17:in `block in _plugins_redmine_boards_watchers_app_views_boards_show_html_erb___1820982835907073717_143660'
/usr/local/bundle/gems/actionview-6.1.7.2/lib/action_view/helpers/capture_helper.rb:45:in `block in capture'
/usr/local/bundle/gems/actionview-6.1.7.2/lib/action_view/helpers/capture_helper.rb:209:in `with_output_buffer'
/usr/local/bundle/gems/actionview-6.1.7.2/lib/action_view/helpers/capture_helper.rb:45:in `capture'
/usr/local/bundle/gems/actionview-6.1.7.2/lib/action_view/helpers/form_helper.rb:450:in `form_for'
/plugins/redmine_boards_watchers/app/views/boards/show.html.erb:14:in `_plugins_redmine_boards_watchers_app_views_boards_show_html_erb___1820982835907073717_143660'
/usr/local/bundle/gems/actionview-6.1.7.2/lib/action_view/base.rb:247:in `public_send'
/usr/local/bundle/gems/actionview-6.1.7.2/lib/action_view/base.rb:247:in `_run'
```

**Comment by Hareramrai** [comment]
@jdehaan, Thanks for reviewing, I have replace reorder_links with reorder_handle based Maeda's suggestion. Now it's again up for code review.

**Comment by jdehaan** [comment]
Ref: https://www.redmine.org/issues/27758

**Comment by jdehaan** [comment]
We are missing the watchers inside the view now here:

http://localhost:8888/redmine/dev/projects/mars-project-mgmt/boards/31

It should look like so on the right side (screenshot from productive server):

image

**Comment by jdehaan** [comment]
We are missing the watchers inside the view now here:

http://localhost:8888/redmine/dev/projects/mars-project-mgmt/boards/31

It should look like so on the right side (screenshot from productive server):

image

**Comment by jdehaan** [comment]
Basically I am following the main [readme](https://github.com/tools-aoeur/redmine_boards_watchers/blob/master/README.textile) of this repository to test the features.

**Comment by Hareramrai** [comment]
> We are missing the watchers inside the view now here:
> 
> http://localhost:8888/redmine/dev/projects/mars-project-mgmt/boards/31
> 
> It should look like so on the right side (screenshot from productive server):
> 
> image

The link was missing because I took the latest code from GitHub, so I have added the links back to the page. Also there is permission override issue. Basically we have remove this [line](https://gestvh01.eu.apn.local/git/PMO-Tooling/redmine_deployment/src/branch/redmine-5.0/src/redmine/lib/redmine/preparation.rb#L150) from `redmine_deployment repo`.

@jdehaan

**Comment by jdehaan** [comment]
Actually it's the reverse happening. The plugin overrides a permission of the core redmine inadvertently.

The permission was introduced here: https://www.redmine.org/issues/3390  and later moved from lib/redmine into a preparation step, after loading plugins to be sure redmine wins.

A better design for the redmine core would ensure the plugin defined permissions are scoped... But these aren't....

A better thing IMO would be to prefix the permissions with some well defined thing and remove that flaw in redmine's design. @swiehr what do you know about this? Maybe in the end these can also be the **same** permission. Basically the watchers plugin allow the same set of operation but in a bulkier way. But then how to hook the feature correctly in so that both normal operation and boards watchers work properly?

**Comment by Hareramrai** [comment]
@jdehaan, I have renamed the permission now it's up for review.

**Review by jdehaan — APPROVED** [review]
👍🏻 Looks good to me

**Pull request merged**